From: Frank Pavlic <pavlic@de.ibm.com>
Subject: fix NULL pointer derefence when using netif_carrier_off
References: 144973

This patch should fix the NULL pointer dereference problem.
When card->dev is still not allocated but a problem occurs 
on one of the three channels causing a device recovery this will
lead to such a kernel panic.

Acked-by: Jan Blunck <jblunck@suse.de>
---
 drivers/s390/net/qeth_main.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

Index: linux-2.6.15/drivers/s390/net/qeth_main.c
===================================================================
--- linux-2.6.15.orig/drivers/s390/net/qeth_main.c
+++ linux-2.6.15/drivers/s390/net/qeth_main.c
@@ -520,7 +520,8 @@ __qeth_set_offline(struct ccwgroup_devic
 	QETH_DBF_TEXT(setup, 3, "setoffl");
 	QETH_DBF_HEX(setup, 3, &card, sizeof(void *));
 	
-	netif_carrier_off(card->dev);
+	if (card->dev && netif_carrier_ok(card->dev))
+		netif_carrier_off(card->dev);
 	recover_flag = card->state;
 	if (qeth_stop_card(card, recovery_mode) == -ERESTARTSYS){
 		PRINT_WARN("Stopping card %s interrupted by user!\n",
@@ -1703,7 +1704,8 @@ qeth_check_ipa_data(struct qeth_card *ca
 					   QETH_CARD_IFNAME(card),
 					   card->info.chpid);
 				card->lan_online = 0;
-				netif_carrier_off(card->dev);
+				if (card->dev && netif_carrier_ok(card->dev))
+					netif_carrier_off(card->dev);
 				return NULL;
 			case IPA_CMD_STARTLAN:
 				PRINT_INFO("Link reestablished on %s "
